server/route.go

```diff
@@ -976,83 +1161,123 @@ func (s *Server) addRoute(c *client, info *Info) (bool, bool) {
-	if hasMax {
+
+	// Create the fast key which will use the subject or 'subject<spc>queue' for queue subscribers.
+	var (
+		_rkey [1024]byte
+		key   []byte
+		qi    int
+	)
+	if sub.queue != nil {
+		// Just make the key subject spc group, e.g. 'foo bar'
+		key = _rkey[:0]
+		key = append(key, sub.subject...)
+		key = append(key, byte(' '))
+		qi = len(key)
+		key = append(key, sub.queue...)
+	} else {
+		key = sub.subject
+	}
+
+	// We always update for a queue subscriber since we need to send our relative weight.
+	var entry *rme
+	var ok bool
+
+	// Always update if a queue subscriber.
+	update := qi > 0
+
+	// Copy to hold outside acc lock.
+	var entryN int32
+
+	acc.mu.Lock()
+	if entry, ok = rm[string(key)]; ok {
+		entry.n += delta
+		if entry.n <= 0 {
+			delete(rm, string(key))
+			update = true // Update for deleting,
+		}
+	} else if delta > 0 {
+		entry = &rme{qi, delta}
+		rm[string(key)] = entry
+		update = true // Adding for normal sub means update.
+	}
+	if entry != nil {
+		entryN = entry.n
+	}
+	acc.mu.Unlock()
+
+	if !update || entry == nil {
 		return
 	}
-	rsid := routeSid(sub)
-	proto := fmt.Sprintf(unsubProto, rsid)
-	s.broadcastInterestToRoutes(sub, proto)
+	// We need to send out this update.
+
+	// If we are sending a queue sub, copy and place in the queue weight.
+	if sub.queue != nil {
+		sub.client.mu.Lock()
+		nsub := *sub
+		sub.client.mu.Unlock()
+		nsub.qw = entryN
+		sub = &nsub
+	}
+
+	// Note that queue unsubs where entry.n > 0 are still
+	// subscribes with a smaller weight.
+	if entryN > 0 {
+		s.broadcastSubscribe(sub)
+	} else {
+		s.broadcastUnSubscribe(sub)
+	}
 }
 
-// Sends UNSUB protocols for each of the subscriptions in the given array
-// to all connected routes. Used when a client connection is closed. Note
-// that when that happens, the subscriptions' MAX have been cleared (force unsub).
-func (s *Server) broadcastUnSubscribeBatch(subs []*subscription) {
-	var (
-		_routes [32]*client
-		routes  = _routes[:0]
-	)
+// broadcastSubscribe will forward a client subscription
+// to all active routes as needed.
+func (s *Server) broadcastSubscribe(sub *subscription) {
+	trace := atomic.LoadInt32(&s.logging.trace) == 1
 	s.mu.Lock()
+	subs := []*subscription{sub}
 	for _, route := range s.routes {
-		routes = append(routes, route)
+		route.mu.Lock()
+		route.sendRouteSubProtos(subs, trace, func(sub *subscription) bool {
+			return route.canImport(sub.subject)
+		})
+		route.mu.Unlock()
 	}
 	s.mu.Unlock()
+}
 
-	for _, route := range routes {
+// broadcastUnSubscribe will forward a client unsubscribe
+// action to all active routes.
+func (s *Server) broadcastUnSubscribe(sub *subscription) {
+	trace := atomic.LoadInt32(&s.logging.trace) == 1
+	s.mu.Lock()
+	subs := []*subscription{sub}
+	for _, route := range s.routes {
 		route.mu.Lock()
-		route.sendRouteUnSubProtos(subs, nil)
+		route.sendRouteUnSubProtos(subs, trace, func(sub *subscription) bool {
+			return route.canImport(sub.subject)
```

Comment:
Not sure we need to check the permissions for the unsubscribe. If we don't, you could pass `nil` for the filter function.

Reply:
Just cuts down on unnecessary traffic IMO. But filters on routes does not make much sense to me. Yes for GWs..

Reply:
Was not thinking straight.. sorry about that.